src/suricata-plugin.h

```diff
@@ -52,4 +52,15 @@ typedef struct SCCapturePlugin_ {
 
 int SCPluginRegisterCapture(SCCapturePlugin *);
 
+typedef struct SCAppLayerPlugin_ {
```

Comment:
needs some struct docs

Reply:
@catenacyber Do you see this as being *only* used for app-layer plugins? Or become the new way to register app-layers? If the latter, it should be more generic than plugin context. Also its a lot odd to be here, but then used in detect-engine-register.c, I think it should be localized to plugin specific code.

Reply:
Just for some context, if "plugin" is mentioned outside of a file with plugin in the name, its probably suspect to a leaky abstraction.  We have this for capture plugins still, but have got rid of it for other types of plugins.

Reply:
> Do you see this as being only used for app-layer plugins?

That is what I thought at first

Reply:
Ok, then I'd move the usage of the `Plugin` struct out of the app-layer.  Keep it within the plugin API, and not the app-layer API. Then register it from there so the plugin interface doesn't leak into the app-layer API if that makes sense.

Reply:
What do you mean ? For example, Which line of code should I move from which file to which file ?

Reply:
I guess this:
```
#ifdef HAVE_PLUGINS
    for (size_t i = 0; i < app_layer_plugins_nb; i++) {
        SCAppLayerPlugin *app_layer_plugin = SCPluginFindAppLayerByIndex(i);
        if (app_layer_plugin == NULL) {
            break;
        }
        app_layer_plugin->Register();
    }
#endif
```
it doesn't really need HAVE_PLUGINS, its simply dynamic app-layer registration right?

It just feels weird that the app-layer "layer" is plugin aware, instead the plugin-layer would be app-layer aware.

Can we get a sample plugin for CI purposes to make sure it works? Might help reason about it.

Reply:
> Can we get a sample plugin for CI purposes to make sure it works? 

Example plugin at https://github.com/catenacyber/suricata-zabbix ;-)

> It just feels weird that the app-layer "layer" is plugin aware, instead the plugin-layer would be app-layer aware.

I see what you mean.
The problem is different initialization timings :
1. plugin is initialized
2. app-layer are registered (and then keywords, and then outputs)

So the plugin init cannot call directly its app-layer registration.
Instead, it registers a callback that will be called during app-layer registration...

Reply:
> Example plugin at https://github.com/catenacyber/suricata-zabbix ;-)

I think we need some minimal in-tree so we can make sure it works.  For example, this is the reason why `examples/plugins/ci-capture/` exists - however we also now have 2 in-tree capture methods that are plugins as well.

Reply:
Done in https://github.com/OISF/suricata/pull/12116